### PR TITLE
*: use pingcap/log (zap) for logging 

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

Switch to use the [uniform log format](https://github.com/tikv/rfcs/blob/master/text/2018-12-19-unified-log-format.md). 

### What is changed and how it works?

The log library is changed from `logrus` to `zap`, a structured logger.

Due to how structured logging usually works, several interfaces are rewritten.

* *Child loggers* are heavily used to abstract away contextual information in the log, e.g. the table name, engine UUID etc. Many types which originally compute these tagging information on-line are changed to storing a child logger instead.
* The 3 functions `QueryRowWithRetry`/`TransactWithRetry`/`ExecWithRetry` originally writes log without any contextual information. These are modified to accept a logger. The common arguments are abstracted into an `SQLWithRetry` type.

All logging infrastructures (e.g. `common.AppLogger`) are moved from `common` into a new `log` package. 

* Like `common.AppLogger`, we need to define our own logger `log.L()` distinct from the true global one `"github.com/pingcap/log".L()` so we could control TiDB's logging level independently (though perhaps we could investigate whether merging the two is possible nowadays). 
* Stack traces of error level logs are entirely disabled, and stack traces of errors are by default inhibited by the `log.ShortError` function, to avoid flooding the log with noise.
* A common pattern used in Lightning is:
     1. Log that a task has started
     2. Perform a task
     3. Log that a task has ended, and display the total time taken.  
  This is codified into the `log.Task` type.

Several redundant error logs are eliminated (e.g. if `f()` calls `g()` calls `h()`, and `h()` logs an error, then `g()` and `f()` will not log the error again). 

Following pingcap/tidb#10025's lead, when refactoring a function, if an error is known to be a custom error, `errors.Trace` will no longer be applied. This will be done more thoroughly in the future, after unit-test coverage is improved.

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Breaking backward compatibility
    * Log format is modified

Related changes

 - Need to update the documentation

### CODE REVIEW

File: lightning/common/util.go [hunk 1]
```diff
@@ -182,35 +182,17 @@ func IsRetryableError(err error) bool {
...
+// affected by whether the error is canceling or not.
 //
 // This function returns `false` (not a context-canceled error) if `err == nil`.
 func IsContextCanceledError(err error) bool {
-	err = errors.Cause(err)
-	return err == context.Canceled || status.Code(err) == codes.Canceled
+	return log.IsContextCanceledError(err)
+}
+
+func ShouldLogError(err error) bool {
```
Comment (kennytm) [inline lightning/common/util.go]: (I'll remove this function later.)

**Comment by kennytm** [comment]
Output looks like:

Screenshot 2019-04-17 at 02 53 56

File: lightning/common/once_error.go [hunk 2]
```diff
@@ -28,16 +28,13 @@ type OnceError struct {
...
 // If this method is called multiple times, only the first call is effective.
-func (oe *OnceError) Set(tag string, e error) {
+func (oe *OnceError) Set(e error) {
 	if e != nil {
 		oe.lock.Lock()
 		if oe.err == nil {
 			oe.err = e
 		}
 		oe.lock.Unlock()
-		if ShouldLogError(e) {
```
Comment (csuzhangxc) [inline lightning/common/once_error.go]: remove `ShouldLogError` in _common/util.go_?

File: lightning/kv/importer.go [hunk 3]
```diff
@@ -102,43 +103,34 @@ func (importer *Importer) SwitchMode(ctx context.Context, mode sst.SwitchMode) e
 			Mode: mode,
 		},
 	}
-	timer := time.Now()
 
+	task := log.With(zap.Stringer("mode", mode)).Begin(zap.DebugLevel, "switch mode")
```
Comment (csuzhangxc) [inline lightning/kv/importer.go]: why do we use `DebugLevel`?

Comment (kennytm) [reply lightning/kv/importer.go]: Originally there was no "switch mode start" log, so this was set to debug level to keep the log from showing by default.

File: lightning/restore/restore.go [hunk 4]
```diff
@@ -1289,38 +1271,26 @@ type RemoteChecksum struct {
...
 		return nil, errors.Trace(err)
 	}
 	// set it back finally
-	defer func() {
-		err = UpdateGCLifeTime(ctx, db, ori)
-		if err != nil && common.ShouldLogError(err) {
-			common.AppLogger.Errorf("[%s] update tikv_gc_life_time error %v", table, errors.ErrorStack(err))
-		}
-	}()
+	defer UpdateGCLifeTime(ctx, db, ori)
```
Comment (csuzhangxc) [inline lightning/restore/restore.go]: log the returned error is better? so we can set it back manually if needed.

File: lightning/log/log.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	case err == nil:
+		level = task.level
+		verb = " completed"
+	case IsContextCanceledError(err):
+		level = zap.DebugLevel
+		verb = " canceled"
+		extraFields = nil
+	default:
+		verb = " failed"
+		extraFields = nil
```
[the remark this answers is not in the document]
Comment (kennytm) [reply lightning/log/log.go]: Yes. Those extra information may be incorrect if the task failed. Added comment.

File: lightning/log/log.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+// level.
+func SetLevel(level zapcore.Level) zapcore.Level {
+	oldLevel := appLevel.Level()
+	appLevel.SetLevel(level)
+	return oldLevel
+}
+
+// ShortError contructs a field which only records the error message without the
+// verbose text (i.e. excludes the stack trace).
+func ShortError(err error) zap.Field {
```
[the remark this answers is not in the document]
Comment (kennytm) [reply lightning/log/log.go]: For Lightning, almost always `ShortError`. No advice for other projects. Will add a comment.

File: lightning/restore/restore.go [hunk 7]
```diff
@@ -481,19 +497,12 @@ func (rc *RestoreController) restoreTables(ctx context.Context) error {
 	for _, dbMeta := range rc.dbMetas {
 		dbInfo, ok := rc.dbInfos[dbMeta.Name]
 		if !ok {
-			common.AppLogger.Errorf("database %s not found in rc.dbInfos", dbMeta.Name)
-			continue
+			return errors.Errorf("database %s not found in rc.dbInfos", dbMeta.Name)
```
[the remark this answers is not in the document]
Comment (kennytm) [reply lightning/restore/restore.go]: Yes unify behavior with L505.

File: lightning/restore/restore.go [hunk 8]
```diff
@@ -817,11 +833,7 @@ func (t *TableRestore) importEngine(
 	if rc.cfg.PostRestore.Level1Compact &&
 		atomic.CompareAndSwapInt32(&rc.compactState, compactStateIdle, compactStateDoing) {
 		go func() {
-			err := rc.doCompact(ctx, Level1Compact)
-			if err != nil {
-				// log it and continue
-				common.AppLogger.Warnf("compact %d failed %v", Level1Compact, err)
-			}
+			rc.doCompact(ctx, Level1Compact)
```
[the remark this answers is not in the document]
Comment (kennytm) [reply lightning/restore/restore.go]: Yes. Added comment.

File: lightning/restore/restore.go [hunk 9]
```diff
@@ -920,9 +925,7 @@ func (rc *RestoreController) switchToNormalMode(ctx context.Context) error {
 }
 
 func (rc *RestoreController) switchTiKVMode(ctx context.Context, mode sstpb.SwitchMode) {
-	if err := rc.importer.SwitchMode(ctx, mode); err != nil {
-		common.AppLogger.Warnf("cannot switch to %s mode: %v", mode.String(), err)
-	}
+	rc.importer.SwitchMode(ctx, mode)
```
[the remark this answers is not in the document]
Comment (kennytm) [reply lightning/restore/restore.go]: ditto

**Comment by GregoryIan** [comment]
@july2993 @lonng PTAL

File: lightning/verification/checksum.go [hunk 10]
```diff
@@ -88,3 +89,10 @@ func (c *KVChecksum) SumSize() uint64 {
 func (c *KVChecksum) SumKVS() uint64 {
 	return c.kvs
 }
+
+func (c *KVChecksum) MarshalLogObject(encoder zapcore.ObjectEncoder) error {
```
Comment (july2993) [inline lightning/verification/checksum.go]: add comment for this? 
ipml zapcore#ObjectMarshaler

File: lightning/restore/restore.go [hunk 11]
```diff
@@ -96,13 +99,17 @@ type errorSummaries struct {
 func (es *errorSummaries) emitLog() {
 	es.Lock()
 	defer es.Unlock()
+
+	logger := log.L()
```
Comment (lonng) [inline lightning/restore/restore.go]: It's OK but better to move this call to if statement.

File: lightning/restore/restore.go [hunk 12]
```diff
@@ -1209,31 +1215,24 @@ func (t *TableRestore) initializeColumns(columns []string, ccp *ChunkCheckpoint)
...
 func (tr *TableRestore) restoreTableMeta(ctx context.Context, db *sql.DB) error {
-	timer := time.Now()
-
-	err := AlterAutoIncrement(ctx, db, tr.tableMeta.DB, tr.tableMeta.Name, tr.alloc.Base()+1)
-	if err != nil {
-		return errors.Trace(err)
-	}
-	common.AppLogger.Infof("[%s] alter table set auto_id takes %v", common.UniqueTable(tr.tableMeta.DB, tr.tableMeta.Name), time.Since(timer))
-	return nil
+	return AlterAutoIncrement(ctx, db, tr.tableName, tr.alloc.Base()+1)
```
Comment (lonng) [inline lightning/restore/restore.go]: I think we can remove this function if only call another function directly.

**Pull request merged**